timemachine/potentials/potentials.py

```diff
@@ -0,0 +1,260 @@
+        )
+
+    def to_gpu(self, precision: Precision) -> GpuImplWrapper:
+        all_pairs = NonbondedAllPairs(
+            self.num_atoms,
+            self.beta,
+            self.cutoff,
+            disable_hilbert_sort=self.disable_hilbert_sort,
+            nblist_padding=self.nblist_padding,
+        )
+        exclusions = NonbondedPairListNegated(self.exclusion_idxs, self.scale_factors, self.beta, self.cutoff)
+        return FanoutSummedPotential([all_pairs, exclusions]).to_gpu(precision)
+
+
+@dataclass
+class NonbondedAllPairs(Potential):
+    num_atoms: int
+    beta: float
+    cutoff: float
+    atom_idxs: Optional[Array] = None
+    disable_hilbert_sort: bool = False
+    nblist_padding: float = 0.1
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        s = self.atom_idxs if self.atom_idxs is not None else slice(None)
+        conf = jnp.array(conf)[s, :]
+        num_atoms, _ = conf.shape
+        no_rescale = jnp.ones((num_atoms, num_atoms))
+
+        return nonbonded.nonbonded(
+            conf,
+            jnp.array(params)[s, :],
+            box,
+            no_rescale,
+            no_rescale,
+            self.beta,
+            self.cutoff,
+            runtime_validate=False,  # needed for this to be JAX-transformable
+        )
+
+
+@dataclass
+class NonbondedInteractionGroup(Potential):
+    num_atoms: int
+    row_atom_idxs: NDArray[np.int32]
+    beta: float
+    cutoff: float
+    disable_hilbert_sort: bool = False
+    nblist_padding: float = 0.1
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        num_atoms, _ = jnp.array(conf).shape
+
+        vdW, electrostatics = nonbonded.nonbonded_interaction_groups(
+            conf,
+            params,
+            box,
+            self.row_atom_idxs,
+            np.setdiff1d(jnp.arange(num_atoms), self.row_atom_idxs),
+            self.beta,
+            self.cutoff,
+        )
+        return jnp.sum(vdW) + jnp.sum(electrostatics)
+
+
+@dataclass
+class NonbondedPairList(Potential):
+    idxs: NDArray[np.int32]
+    rescale_mask: NDArray[np.float64]
+    beta: float
+    cutoff: float
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        vdW, electrostatics = nonbonded.nonbonded_on_specific_pairs(
+            conf, params, box, self.idxs, self.beta, self.cutoff, self.rescale_mask
+        )
+        return jnp.sum(vdW) + jnp.sum(electrostatics)
+
+
+@dataclass
+class NonbondedPairListNegated(Potential):
+    idxs: NDArray[np.int32]
+    rescale_mask: NDArray[np.float64]
+    beta: float
+    cutoff: float
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        vdW, electrostatics = nonbonded.nonbonded_on_specific_pairs(
+            conf, params, box, self.idxs, self.beta, self.cutoff, self.rescale_mask
+        )
+        return -jnp.sum(vdW) - jnp.sum(electrostatics)
```

Comment:
Opinionated super-nit: apply negation once `U = jnp.sum(vdW) + jnp.sum(electrostatics); return -U`

Reply:
I prefer your suggestion too. Willfix.

Reply:
Changed in https://github.com/proteneer/timemachine/pull/995/commits/df34ffeea9136d3d7081ac472f59407a492c03ae